### PR TITLE
enable if the dot product kernel

### PR DESCRIPTION
**Pull request by drezap**
## Summary

Add `enable_if` to the two different specializations of the dot product kernel, i.e. `vector x[N]` and vector[D] x[N]` so that the dot product will compile with command Stan.

## Tests

Passes all the same tests.

## Side Effects

No side effects.

## Checklist

- [x] Math issue #979

- [x] Copyright holder: Andre Zapico

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by drezap** [comment]
This is failing because of the same issue #965. Once that one's in, I can rebase with dev and this will be fine.

**Comment by bob-carpenter** [comment]
I'm not sure what you mean by rebase with dev, but please don't rewrite history on the GitHub origin.

**Comment by drezap** [comment]
I just mean to pull from upstream develop, so that the updated are included in this PR. Sorry - did not know the meaning of rebase.

**Comment by bbbales2** [comment]
This needs an issue. Did functionality change? It looks like that might be the case cause all the input validation tests are changing. Is there a roadmap somewhere for this GP development stuff, by the way? It might help to have one if not.

Discourse is a fine place for something high level like that. You can reference it in your issues or whatever, instead of having to re-explain the basics all the time.

**Comment by drezap** [comment]
@bbbales2 issue is #979. Functionality did not change, so all of the tests remain the same. I'm sorry - I don't know what you mean by "all the validation tests are changing".

I think I actually just forked this off the wrong branch, and I should have done develop, and that's why the tests are failing.

There's no roadmap, only for "Developing a Gaussian Process Covariance Function", but I'm happy to put something together. I can open up a post and we can discuss what's important w this library moving forward.

**Comment by bbbales2** [comment]
Whoops, my bad. I see the issue, sorry.

So are these changes: https://github.com/stan-dev/math/pull/980/files#diff-933b2d4c981eea94d73b064655840912R119

not part of this pull? If so, maybe merge in the latest master and Github won't show them?

**Comment by bbbales2** [comment]
With respect to roadmap, I think that'd be good. I know this is something we want to start doing for sizable development efforts (and what you're working on is certainly sizable).

The problem here is that I'm not super familiar with how to write such a document. What length, what contents, etc. Lemme just open a Discourse thread and @ a couple people and we'll see what to do.

**Comment by drezap** [comment]
@bbbales2 I'm not sure why I can't see these changes. I simply added some `boost::enable_if` statements, that check whether `T_x` `is_vector_like` or `is_scalar`. Can you see these changes? I can click on "files changed" and I see them just fine.

Please let me know if there's any trouble. There was only a few lines of changed code.

**Comment by bbbales2** [comment]
I just took a screenshot of my browser:

screenshot_diff

The check_finite moved, it seems.

**Comment by drezap** [comment]
Yes, I edited that, too, apparently. but more important is the templating at the function signature, at lines 44 through 47. Please check that. I forgot I made these changes. These are just changes I made to make sure this worked in command stan, and I could use it in stan models.

**Comment by drezap** [comment]
enable_if_dot_prod

**Comment by drezap** [comment]
this should be evident from the "files changed" tab

**Comment by drezap** [comment]
Are you requiring any changes? If so, please let me know what need be changed.

**Comment by bbbales2** [comment]
Why does the loop here: https://github.com/stan-dev/math/blob/0a6d990596a272c7f753fe557583d37e46ed06ba/stan/math/prim/mat/fun/gp_dot_prod_cov.hpp#L72 go to size - 1?

Line 79 only fills in the diagonal

**Comment by bob-carpenter** [comment]
The loop covers 
```
0 <= i < x.size() - 1
```
and 
```
i + 1 <= j < .size()
```
and then sets `(i, j)` and `(j, i)`.  The only index this misses is `(x.size() - 1, x.size() - 1)`.

**Comment by bbbales2** [comment]
Review ahoy! I talked to @bob-carpenter a little about the enable_ifs, and you can avoid them here.

Option 1, use the signatures:
`gp_dot_prod_cov(const std::vector<T_x> &x, const T_sigma &sigma)`
and
`gp_dot_prod_cov(const std::vector<Eigen::Matrix<T_x, Eigen::Dynamic, 1>> &x, const T_sigma &sigma)`

That way the argument lookup determines which function to use, rather than the enable_ifs.

Option 2, you should be able to compress these 4 functions into two by just going with the signature
`gp_dot_prod_cov(const std::vector<T_x> &x, const T_sigma &sigma)`

And then writing a helper function in a private namespace to handle the dot product bit. Something like:
```
namespace {
  double gp_dot_prod_fun(double a1, double a2) {
     return a * a2;
  }

  double gp_dot_prod_fun(Eigen::VectorXd a1, Eigen::VectorXd a2) {
    return a1.transpose() * a2;
  }
}
```

would be the double implementation.

I think the code:

```
for (size_t i = 0; i < x_size; ++i) {
    check_not_nan("gp_dot_prod_cov", "x", x[i]);
    check_finite("gp_dot_prod_cov", "x", x[i]);
}
```

Should work whether x[i] is a scalar or an `Eigen::Matrix<T, Eigen::Dynamic, 1>`. Can you make that change whichever Option you end up deciding on?

Also re: `size_t D = length(x[0]); // assuming length all the same length`, since you're looping through each input, go ahead and be explicit about checking I guess. I think it's a valid assumption in Stan though since we don't have ragged arrays.

**Comment by drezap** [comment]
@bbbales2 .... so I'm working on option 1, and I'm not able to get this to work quickly. This also doesn't consider the fact that there could be row vector and vector inputs. The `enable_if`'s take care of both of these, and there's no performance gain. 

Option 2 I could get to work with some hours of labor, and it might make the code more "idiomatic".

There's also no active review on this PR. Is this _required_ that I include these changes? or are these suggestions?

Hi - I'm trying to be nice while saying this: I have limited time to spend on these before my semester starts, and I'm finding I'm changing code that works fine, with no performance gain. I appreciate the suggestions, but we can't be so pedantic that nothing gets done. We're spending time discussing minor details and nothing is getting done. 

Please open a formal review, and **tell me what is required** to make this PR go through.

**Comment by bbbales2** [comment]
> tell me what is required

The issue with this is it's been difficult for me to figure out what the problem is. There's this issue https://github.com/stan-dev/math/issues/979, but I'm not sure what it means, so I've been trying to dig into the code to figure out what's going on.

Your question was, what do you need to finish this pull request.

1. This is a bug of some sort, right?

From the pull req template (https://github.com/stan-dev/math/blob/develop/.github/PULL_REQUEST_TEMPLATE.md), bug pull requests should come with at least one test that would make it impossible to re-implement this bug that apparently snuck through the first code review.

So that should definitely be in there. My bad for missing it the first few looks.

2. enable_ifs are a weird workaround for handling different input types. We want to avoid them where we can. Idiomatic is the word. It's annoying, but whatevs.

Since you need to handle row vectors, how about the signatures:
```
gp_dot_prod_cov(const std::vector<T_x> &x, const T_sigma &sigma)
gp_dot_prod_cov(const std::vector<Eigen::Matrix<T_x, R, C>> &x, const T_sigma &sigma)
```

where R and C are template parameters? You could check at runtime that one of R and C was 1. There's probably an error program that does that, but I don't think it would compile anyway cause there probably isn't a matrix input version of dot_product.

3. Those error checks should be vectorized.

```
for (size_t i = 0; i < x_size; ++i) {
    check_not_nan("gp_dot_prod_cov", "x", x[i]);
    check_finite("gp_dot_prod_cov", "x", x[i]);
}
```

should work when the input is a std::vector of Eigen::VectorXds or Eigen::RowVectorXds. If they don't, my bad, but I did check the implementation and I think this will work.

4. I'm trying hard to not get out of the scope of this pull request, but these tparam comments are questionable:

` * @tparam T_x type of std::vector of elements` and `* @tparam T_x type of std::vector of double`

How about

` * @tparam T_x type of elements of x`

for when T_x is an Eigen::Matrix, or for the cases where the inputs are scalars

` * @tparam T_x scalar type of x`

I promise I'm trying not to read this too closely haha, but if I notice something I'm gonna call it out.

File: stan/math/prim/mat/fun/gp_dot_prod_cov.hpp [hunk 1]
```diff
@@ -47,10 +53,13 @@ gp_dot_prod_cov(const std::vector<T_x> &x, const T_sigma &sigma) {
...
 
   size_t x_size = x.size();
-  for (size_t i = 0; i < x_size; ++i)
-    check_not_nan("gp_dot_prod_cov", "x", x[i]);
-
-  check_finite("gp_dot_prod_cov", "x", x);
+  size_t D = x[0].size();
+  for (size_t i = 0; i < x_size; ++i) {
+    for (size_t d = 0; d < D; ++d) {
+      check_not_nan("gp_dot_prod_cov", "x", x[i][d]);
```
Comment (bbbales2) [inline stan/math/prim/mat/fun/gp_dot_prod_cov.hpp]: These checks can be vectorized across the D loop.

File: stan/math/prim/mat/fun/gp_dot_prod_cov.hpp [hunk 2]
```diff
@@ -93,10 +102,12 @@ gp_dot_prod_cov(const std::vector<T_x> &x, const T_sigma &sigma) {
  * @throw std::domain_error if sigma < 0, nan, inf or
  *   x is nan or infinite
  */
-template <typename T_sigma>
-inline typename Eigen::Matrix<typename stan::return_type<double, T_sigma>::type,
-                              Eigen::Dynamic, Eigen::Dynamic>
-gp_dot_prod_cov(const std::vector<double> &x, const T_sigma &sigma) {
+template <typename T_x, typename T_sigma>
+inline typename boost::enable_if_c<
+    is_constant<T_x>::value,
```
Comment (bbbales2) [inline stan/math/prim/mat/fun/gp_dot_prod_cov.hpp]: Change the function signature so it returns a matrix. Like:

```
gp_dot_prod_cov(const std::vector<Eigen::Matrix<T_x, Eigen::Dynamic, 1>> &x, const T_sigma &sigma)
```

Let's not worry about ```Eigen::Matrix<T_x, 1, Eigen::Dynamic>``` types

File: stan/math/prim/mat/fun/gp_dot_prod_cov.hpp [hunk 3]
```diff
@@ -166,18 +176,21 @@ gp_dot_prod_cov(const std::vector<T_x1> &x1, const std::vector<T_x2> &x2,
...
-    check_not_nan("gp_dot_prod_cov", "x2", x2[i]);
-
-  check_finite("gp_dot_prod_cov", "x2", x2);
-
-  Eigen::Matrix<typename stan::return_type<T_x1, T_x2, T_sigma>::type,
-                Eigen::Dynamic, Eigen::Dynamic>
+  size_t D = x1[0].size();
+  for (size_t i = 0; i < x1_size; ++i) {
+    for (size_t d = 0; d < D; ++d) {
+      check_not_nan("gp_dot_prod_cov", "x1", x1[i][d]);
```
Comment (bbbales2) [inline stan/math/prim/mat/fun/gp_dot_prod_cov.hpp]: Vectorize

File: stan/math/prim/mat/fun/gp_dot_prod_cov.hpp [hunk 4]
```diff
@@ -166,18 +176,21 @@ gp_dot_prod_cov(const std::vector<T_x1> &x1, const std::vector<T_x2> &x2,
...
+      check_finite("gp_dot_prod_cov", "x1", x1[i][d]);
+    }
+  }
+  for (size_t i = 0; i < x2_size; ++i) {
+    for (size_t d = 0; d < D; ++d) {
+      check_not_nan("gp_dot_prod_cov", "x2", x2[i][d]);
```
Comment (bbbales2) [inline stan/math/prim/mat/fun/gp_dot_prod_cov.hpp]: Vectorize

File: stan/math/prim/mat/fun/gp_dot_prod_cov.hpp [hunk 5]
```diff
@@ -2,12 +2,16 @@
...
 #include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/dot_product.hpp>
-#include <stan/math/prim/mat/fun/value_of.hpp>
 #include <stan/math/prim/scal/err/check_finite.hpp>
 #include <stan/math/prim/scal/err/check_nonnegative.hpp>
 #include <stan/math/prim/scal/err/check_not_nan.hpp>
 #include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/mat/fun/dot_product.hpp>
+#include <stan/math/prim/mat/fun/dot_self.hpp>
+#include <boost/utility/enable_if.hpp>
```
Comment (bbbales2) [inline stan/math/prim/mat/fun/gp_dot_prod_cov.hpp]: Once you change the function signatures, you can get rid of some of these new headers.

Comment (drezap) [reply stan/math/prim/mat/fun/gp_dot_prod_cov.hpp]: Hey - I still need these two function signatures. There are two cases: a 1D GP, where the dot product is element wise multiplication (dot_self). D>1 => I need dot_product... is this clear enough?

**Comment by drezap** [comment]
@bbbales2  hey do you have a preferred way of "dev"ing stuff that's exposed to users, so I don't have this issue again? run it out out command stan and test as we go? is there a test folder I can expose functions to users to make sure I'm templating the stuff properly? or how can I write unit tests accordingly?

File: stan/math/prim/mat/fun/gp_dot_prod_cov.hpp [hunk 6]
```diff
@@ -1,14 +1,18 @@
 #ifndef STAN_MATH_PRIM_MAT_FUN_COV_DOT_PROD_HPP
 #define STAN_MATH_PRIM_MAT_FUN_COV_DOT_PROD_HPP
 
+#include <boost/utility/enable_if.hpp>
```
Comment (bbbales2) [inline stan/math/prim/mat/fun/gp_dot_prod_cov.hpp]: Not used, so get rid of it. I don't think is_vector_like, check_nonnegative, or is_constant are used either.

**Comment by bbbales2** [comment]
Re: catching mistakes, nah, I don't have any great advice. It's just gonna happen (https://github.com/stan-dev/math/pull/1023).

It's always worth being more careful, but Bob & I both reviewed the pull and didn't catch it or the missing tests and I assume we were both trying to be careful.

**Comment by drezap** [comment]
@bbbales2 issues resolved

**Comment by bbbales2** [comment]
Cool. I think the right test is actually there already now that I check. I'll merge when test passes.

**Pull request merged**